src/input_utils.py

```diff
@@ -315,17 +315,241 @@ def add_side_input_features(
+  mask_token_id = vocab.index(b'[MASK]')
+
+  text_item_selector = tf_text.RandomItemSelector(
+      max_selections_per_batch=mlm_max_selections_per_sequence,
+      selection_rate=mlm_fraction_to_mask,
+      unselectable_ids=unselectable_ids)
+
+  patch_item_selector = tf_text.RandomItemSelector(
+      max_selections_per_batch=mpp_max_selections_per_sequence,
+      selection_rate=mpp_fraction_to_mask,
+      unselectable_ids=unselectable_ids)
+
+  mask_values_chooser = tf_text.MaskValuesChooser(len(vocab),
+                                                  mask_token_id, 0.8)
+  num_patch_per_row = image_size // patch_size
+  num_patches = num_patch_per_row ** 2
+
+  def make_masked_patch_label_ids(masked_patch_embeddings):
+    """Makes taget labels for masked patch prediction.
+  
+    Args:
+      masked_patch_embeddings:
+        <tf.Tensor>[batch, sequence_length, patch_embedding_size].
+  
+    Returns:
+      <tf.Tensor>[batch, sequence_length].
+
+    """
+
+    bin_size = max_pixel_val // (2 ** output_channel_bits)
+  
+    # Scale from 0-1 to 0-255
+    masked_patch_embeddings = masked_patch_embeddings * (max_pixel_val - 1)
+    avg_target = reduce(
+        masked_patch_embeddings,
+        'batch seq_len (p2 channels) -> batch (seq_len channels)',
+        'mean',
+        channels=channels,
+        p2=patch_size**2)
+
+    channel_bins = list(range(bin_size, max_pixel_val, bin_size))
+
+    discretized_target = math_ops._bucketize(avg_target, channel_bins)
+    discretized_target = tf.cast(discretized_target, tf.int32)
+    discretized_target = rearrange(
+        discretized_target,
+        'batch (seq_len channels) -> batch seq_len channels',
+        channels=channels)
+    
+    bin_mask = (2 ** output_channel_bits) ** tf.range(0, channels)
+    bin_mask = rearrange(bin_mask, 'channels -> () () channels')
+    label_ids = bin_mask * discretized_target
+    label_ids = reduce(
+        label_ids,
+        'batch seq_len channels -> batch seq_len',
+        'sum')
+    return label_ids
+
+  def get_masked_weights(masked_token_ids: tf.Tensor,
+                         masked_seq_len: int) -> tf.Tensor:
+    """Gets mask for real predictions.
+
+    The `positions` tensor might be zero-padded (if the sequence is too short 
+    to have the maximum number of predictions).
+    The `masked_weights` tensor has a value of 1.0 for every real prediction 
+    and 0.0 for the padding predictions.
+
+    Args:
+      masked_token_ids: A tensor that contains mask token ids.
+    Returns:
+      A mask that indicate the positions of real predictions.
+
+    """
+    mask_position = tf.equal(masked_token_ids, mask_token_id)
+    mask_position = tf.cast(mask_position, tf.int32)
+    num_real_masked_tokens = tf.reduce_sum(mask_position, axis=1, keepdims=True)
+    position = tf.range(masked_seq_len, dtype=tf.int32)
+    return tf.where(position < num_real_masked_tokens, 1.0, 0.0)
+    
+  def pad_to_max_seq_len(tensor, max_seq_len):
```

Comment:
What happens if max_seq_len is actually smaller than the length of tensor?

Reply:
If we call [get_decode_fn](https://github.com/googleinterns/multimodal-long-transformer-2021/blob/roylu-input/src/input_utils.py#L93) first, we will trim the sequence to be shorter than maximum sequence length. We set make sure that [text part will be shorter than max_text_seq_len](https://github.com/googleinterns/multimodal-long-transformer-2021/blob/roylu-input/src/input_utils.py#L135) and patch part is always 198 (196 + 2). Thus, we will never have a tensor longer than `max_seq_len`.

Reply:
ok! If you have a chance, add a comment somewhere to make sure reading the code it's obvious where trimming occurs :)